File: Term2/Homework2/Kruskal/lib/Kruskal.cs
namespace Kruskal;

using System.Collections.Generic;

public class Kruskal
{
	private List<Edge>[] graph;
	private List<Edge> edgeList;
	private int n;

	public Kruskal(int[,] graph)
	{
		n = graph.GetLength(0);

		if (n != graph.GetLength(1)) 
		{
			throw new ArgumentException("Graph should be given as a square matrix");
		}

		this.graph = new List<Edge>[n];
		this.edgeList = new List<Edge>();

		for (int i = 0; i < n; i++) 
		{
			this.graph[i] = new List<Edge>();
			for (int j = 0; j < n; j++) 
			{
				if (graph[i, j] == -1)
					continue;

				this.edgeList.Add(new Edge(i, j, graph[i, j]));
				this.graph[i].Add(new Edge(i, j, graph[i, j]));
			}
		}
	}

	public int FindMSTWeight() 
	{
		DSU dsu = new DSU(n);

		edgeList.Sort();

		int resultWeight = 0;
		foreach (var edge in edgeList) 
		{
			if (!dsu.JoinSets(edge.From, edge.To))
				resultWeight += edge.Weight;
		}

		return resultWeight;
	}

	public int[,] FindMST() 
	{
		int[,] result = new int[n, n];
		for (int i = 0; i < n; i++)
		{
			for (int j = 0; j < n; j++)
			{
				result[i, j] = -1;
			}
		}

		DSU dsu = new DSU(n);

		edgeList.Sort();

		foreach (var edge in edgeList) 
		{
			if (!dsu.JoinSets(edge.From, edge.To))
				result[edge.From, edge.To] = result[edge.To, edge.From] = edge.Weight;
		}
		return result;
	}
}

File: Term2/Homework2/Kruskal/src/main/java/Kruskal/Kruskal.java
package Kruskal;

import DSU.DSU;
import utils.Edge;

import java.util.ArrayList;
import java.util.Arrays;

public class Kruskal {
	protected ArrayList<Edge>[] graph;
	protected ArrayList<Edge> edgeList;
	private int n;

	public Kruskal(ArrayList<Edge>[] graph) {
		n = graph.length;
		this.graph = graph;
		this.edgeList = new ArrayList<>();

		for (var vertice : graph) {
			edgeList.addAll(vertice);
		}
	}

	private ArrayList<Edge>[] createEmptyGraph(int n) {
		ArrayList<Edge>[] result = (ArrayList<Edge>[]) new ArrayList[n];
		for (int i = 0; i < n; i++)
			result[i] = new ArrayList<>();

		return result;
	}

	private boolean isConnected(DSU dsu) {
		for (int i = 0; i < n; i++) {
			if (dsu.getParent(i) != dsu.getParent(0))
				return false;
		}

		return true;
	}

	public ArrayList<Edge>[] getMST() {
		DSU dsu = new DSU(n);

		var mstGraph = createEmptyGraph(n);

		edgeList.sort(Edge::compareTo);
		for (var edge : edgeList) {
			if (!dsu.joinSets(edge.getFrom(), edge.getTo())) {
				mstGraph[edge.getFrom()].add(edge);
				mstGraph[edge.getTo()].add(edge.getReverseEdge());
			}
		}

		if (!isConnected(dsu)) {
			throw new IllegalArgumentException("Given graph isn't connected");
		}

		return mstGraph;
	}

	public int getMSTWeight() {
		var mstGraph = getMST();
		int resultWeight = 0;

		for (var vertice : mstGraph) {
			for (var edge : vertice)
				resultWeight += edge.getWeight();
		}

		return resultWeight / 2;
	}
}
